independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/jaxrs/ResponseImpl.java

```diff
@@ -283,22 +283,22 @@ public Builder getLinkBuilder(String relation) {
 
     @Override
     public MultivaluedMap<String, String> getStringHeaders() {
-        // FIXME: is this mutable?
         if (stringHeaders == null) {
             // let's keep this map case-insensitive
             stringHeaders = new CaseInsensitiveMap<>();
-            for (Entry<String, List<Object>> entry : headers.entrySet()) {
-                List<String> stringValues = new ArrayList<>(entry.getValue().size());
-                for (Object value : entry.getValue()) {
-                    stringValues.add(HeaderUtil.headerToString(value));
-                }
-                stringHeaders.put(entry.getKey(), stringValues);
-            }
+            headers.forEach(this::populateStringHeaders);
```

Comment:
I don't think it's actually a good idea in this case, as the `forEach` implementation being invoked here is not the one from HashMap which actually implements the optimisaation I referred to.

I'd suggest to

1. refactor the code futher to make the `headers` type of explicitly one type only (`CaseInsensitiveMap` should work?)
2. then try to implement the better `forEach` for the custom implementation it's actually going to use (`CaseInsensitiveMap`).

otherwise, we can still take the patch as I don't think it hurts but it's a bit pointless to change

Reply:
Actually it's the optimization that `TreeMap` uses because `forEach` of `MultivaluedTreeMap` (of which `CaseInsensitiveMap` is a sublass) now delegates to `forEach` of the backing map which is a `TreeMap`.

Does that make sense?

Reply:
ah awesome, I missed that you added the delegate :+1: